lib/Data/Filesystem.php

```diff
@@ -223,6 +247,97 @@ public function existsComment($pasteid, $parentid, $commentid)
         );
     }
 
+    /**
+     * Save a value.
+     *
+     * @access public
+     * @param  string $value
+     * @param  string $namespace
+     * @param  string $key
+     * @return bool
+     */
+    public function setValue($value, $namespace, $key = '')
+    {
+        switch ($namespace) {
+            case 'purge_limiter':
+                return self::_storeString(
+                    self::$_path . DIRECTORY_SEPARATOR . 'purge_limiter.php',
+                    '<?php' . PHP_EOL . '$GLOBALS[\'purge_limiter\'] = ' . $value . ';'
+                );
+            case 'salt':
+                return self::_storeString(
+                    self::$_path . DIRECTORY_SEPARATOR . 'salt.php',
+                    '<?php # |' . $value . '|'
+                );
+            case 'traffic_limiter':
+                self::$_last_cache[$key] = $value;
+                return self::_storeString(
+                    self::$_path . DIRECTORY_SEPARATOR . 'traffic_limiter.php',
+                    '<?php' . PHP_EOL . '$GLOBALS[\'traffic_limiter\'] = ' . var_export(self::$_last_cache, true) . ';'
+                );
```

Comment:
What about a `default` blcok that throws an exception or so? – if the function is called with invalid parameters, IMHOP´ we can panick.

Or at least log an error…

This is to guard against improper usage of the functions given that PHP's type system is quite limited.

Reply:
It is really a code style question if we use exceptions or defined error values for error signaling (empty string, in this case - all storage operations are expected to return a non-empty value). All of the existing interfaces avoided exceptions, so I stuck to the same convention for the new methods.

Of course, I agree that errors do need to be dealt with, regardless if they are returned as `false`, an dedicated error value (empty string, -1), throw an exception or a result-type that either contains an error or the result (like go or rust do, for example).

So, do we handle the error condition at the moment? If the paste can't be read, we return false and this (you'll love the irony) raises an exception, caught in the controller:
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Model/Paste.php#L36-L38
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Controller.php#L317-L327

But the traffic limiter doesn't:
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Persistence/TrafficLimiter.php#L175
I'm a bit on the fence, in this case - we should consider a logging system as this is a failure that should be logged, but doesn't prevent the user from creating the paste (we could read the file but not save it). The same goes for the other two persistence classes:
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Persistence/ServerSalt.php#L68
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Persistence/PurgeLimiter.php#L76

I think I saw the use of an error log mechanism in the GCS class that we could reuse for these cases, as a starting point?

Reply:
I added an error_log() call for these three cases, so errors storing these, after being read, can be found by the server operator in the php logs (they are not getting translated, as that would be the language of the visitor). In the case of the purge limiter, we skip the purge cycle, as in that case the limit would be likely outdated and potentially cause a purge to be triggered every time a paste gets created.

There are currently no unit tests covering that line - it could probably be exercised using the file storage and making the files read-only or such, after creating them.